File: tb3/cliente-python/cliente.py
#!/usr/bin/env python3
"""
Cliente Python para API REST do Sistema de Telefonia
Trabalho 3 - Sistemas Distribuídos
"""

import requests
import json
from typing import Dict, List, Optional

class ClienteTelefoniaAPI:
    """Cliente para consumir a API REST do sistema de telefonia"""
    
    def __init__(self, base_url: str = "http://localhost:8080"):
        self.base_url = base_url
        self.session = requests.Session()
        self.session.headers.update({
            'Content-Type': 'application/json',
            'Accept': 'application/json'
        })
    
    def verificar_servidor(self) -> bool:
        """Verifica se o servidor está online"""
        try:
            response = self.session.get(f"{self.base_url}/api/health", timeout=2)
            return response.status_code == 200
        except:
            return False
    
    def adicionar_cliente(self, nome: str, cpf: str, telefone: str, email: str) -> str:
        """Adiciona um novo cliente"""
        dados = {
            "nome": nome,
            "cpf": cpf,
            "telefone": telefone,
            "email": email
        }
        response = self.session.post(f"{self.base_url}/api/clientes", json=dados)
        return response.json().get("message", "Erro")
    
    def listar_clientes(self) -> List[str]:
        """Lista todos os clientes"""
        response = self.session.get(f"{self.base_url}/api/clientes")
        return response.json()
    
    def consultar_cliente(self, cpf: str) -> str:
        """Consulta informações de um cliente"""
        response = self.session.get(f"{self.base_url}/api/clientes/{cpf}")
        return response.json().get("message", "Erro")
    
    def remover_cliente(self, cpf: str) -> str:
        """Remove um cliente"""
        response = self.session.delete(f"{self.base_url}/api/clientes/{cpf}")
        return response.json().get("message", "Erro")
    
    def adicionar_linha(self, cpf: str, numero: str) -> str:
        """Adiciona uma linha telefônica"""
        dados = {"cpf": cpf, "numero": numero}
        response = self.session.post(f"{self.base_url}/api/linhas", json=dados)
        return response.json().get("message", "Erro")
    
    def remover_linha(self, cpf: str, numero: str) -> str:
        """Remove uma linha telefônica"""
        dados = {"cpf": cpf, "numero": numero}
        response = self.session.request("DELETE", f"{self.base_url}/api/linhas", json=dados)
        return response.json().get("message", "Erro")
    
    def registrar_chamada(self, origem: str, destino: str, duracao: int) -> str:
        """Registra uma chamada"""
        dados = {"origem": origem, "destino": destino, "duracao": duracao}
        response = self.session.post(f"{self.base_url}/api/chamadas", json=dados)
        return response.json().get("message", "Erro")
    
    def gerar_fatura(self, cpf: str) -> str:
        """Gera fatura de um cliente"""
        dados = {"cpf": cpf}
        response = self.session.post(f"{self.base_url}/api/faturas", json=dados)
        return response.json().get("message", "Erro")
    
    def listar_faturas(self, cpf: str) -> List[str]:
        """Lista faturas de um cliente"""
        response = self.session.get(f"{self.base_url}/api/faturas/{cpf}")
        return response.json()
    
    def obter_estatisticas(self) -> str:
        """Obtém estatísticas do sistema"""
        response = self.session.get(f"{self.base_url}/api/estatisticas")
        return response.json().get("message", "Erro")


def exibir_menu():
    """Exibe o menu principal"""
    print("\n╔════════════════════════════════════╗")
    print("║      MENU PRINCIPAL (PYTHON)       ║")
    print("╠════════════════════════════════════╣")
    print("║  1 - Adicionar Cliente             ║")
    print("║  2 - Remover Cliente               ║")
    print("║  3 - Consultar Cliente             ║")
    print("║  4 - Listar Clientes               ║")
    print("║  5 - Adicionar Linha               ║")
    print("║  6 - Remover Linha                 ║")
    print("║  7 - Registrar Chamada             ║")
    print("║  8 - Gerar Fatura                  ║")
    print("║  9 - Listar Faturas                ║")
    print("║ 10 - Ver Estatísticas              ║")
    print("║  0 - Sair                          ║")
    print("╚════════════════════════════════════╝")


def main():
    print("╔════════════════════════════════════════════════════════╗")
    print("║    CLIENTE PYTHON - SISTEMA DE TELEFONIA REST API      ║")
    print("╚════════════════════════════════════════════════════════╝")
    
    api = ClienteTelefoniaAPI()
    
    # Verificar conexão
    print("\n🔌 Verificando conexão com servidor...")
    if not api.verificar_servidor():
        print("❌ Erro: Servidor não está respondendo!")
        print("💡 Certifique-se de que o servidor Java está rodando na porta 8080")
        return
    
    print("✅ Conectado ao servidor!")
    
    while True:
        try:
            exibir_menu()
            opcao = input("Opção: ").strip()
            
            if opcao == "1":
                print("\n=== ADICIONAR CLIENTE ===")
                nome = input("Nome: ")
                cpf = input("CPF: ")
                telefone = input("Telefone: ")
                email = input("Email: ")
                print("\n📌", api.adicionar_cliente(nome, cpf, telefone, email))
            
            elif opcao == "2":
                print("\n=== REMOVER CLIENTE ===")
                cpf = input("CPF: ")
                print("\n📌", api.remover_cliente(cpf))
            
            elif opcao == "3":
                print("\n=== CONSULTAR CLIENTE ===")
                cpf = input("CPF: ")
                print("\n", api.consultar_cliente(cpf))
            
            elif opcao == "4":
                print("\n=== LISTA DE CLIENTES ===")
                clientes = api.listar_clientes()
                if clientes:
                    for i, cliente in enumerate(clientes, 1):
                        print(f"{i}. {cliente}")
                else:
                    print("Nenhum cliente cadastrado")
            
            elif opcao == "5":
                print("\n=== ADICIONAR LINHA ===")
                cpf = input("CPF do cliente: ")
                numero = input("Número da linha: ")
                print("\n📌", api.adicionar_linha(cpf, numero))
            
            elif opcao == "6":
                print("\n=== REMOVER LINHA ===")
                cpf = input("CPF do cliente: ")
                numero = input("Número da linha: ")
                print("\n📌", api.remover_linha(cpf, numero))
            
            elif opcao == "7":
                print("\n=== REGISTRAR CHAMADA ===")
                origem = input("Número de origem: ")
                destino = input("Número de destino: ")
                duracao = int(input("Duração (minutos): "))
                print("\n📌", api.registrar_chamada(origem, destino, duracao))
            
            elif opcao == "8":
                print("\n=== GERAR FATURA ===")
                cpf = input("CPF do cliente: ")
                print("\n", api.gerar_fatura(cpf))
            
            elif opcao == "9":
                print("\n=== LISTAR FATURAS ===")
                cpf = input("CPF do cliente: ")
                faturas = api.listar_faturas(cpf)
                print()
                for fatura in faturas:
                    print(fatura)
            
            elif opcao == "10":
                print("\n=== ESTATÍSTICAS DO SISTEMA ===")
                print("\n", api.obter_estatisticas())
            
            elif opcao == "0":
                print("\n👋 Encerrando cliente Python...")
                break
            
            else:
                print("❌ Opção inválida!")
        
        except KeyboardInterrupt:
            print("\n\n👋 Encerrando...")
            break
        except Exception as e:
            print(f"❌ Erro: {e}")


if __name__ == "__main__":
    main()

File: tb3/cliente-nodejs/cliente.js
/**
 * Cliente Node.js para API REST do Sistema de Telefonia
 * Trabalho 3 - Sistemas Distribuídos
 */

const axios = require('axios');
const readline = require('readline');

class ClienteTelefoniaAPI {
    constructor(baseURL = 'http://localhost:8080') {
        this.api = axios.create({
            baseURL: baseURL,
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json'
            },
            timeout: 5000
        });
    }

    async verificarServidor() {
        try {
            const response = await this.api.get('/api/health');
            return response.status === 200;
        } catch {
            return false;
        }
    }

    async adicionarCliente(nome, cpf, telefone, email) {
        const response = await this.api.post('/api/clientes', { nome, cpf, telefone, email });
        return response.data.message;
    }

    async listarClientes() {
        const response = await this.api.get('/api/clientes');
        return response.data;
    }

    async consultarCliente(cpf) {
        const response = await this.api.get(`/api/clientes/${cpf}`);
        return response.data.message;
    }

    async removerCliente(cpf) {
        const response = await this.api.delete(`/api/clientes/${cpf}`);
        return response.data.message;
    }

    async adicionarLinha(cpf, numero) {
        const response = await this.api.post('/api/linhas', { cpf, numero });
        return response.data.message;
    }

    async removerLinha(cpf, numero) {
        const response = await this.api.delete('/api/linhas', { data: { cpf, numero } });
        return response.data.message;
    }

    async registrarChamada(origem, destino, duracao) {
        const response = await this.api.post('/api/chamadas', { origem, destino, duracao });
        return response.data.message;
    }

    async gerarFatura(cpf) {
        const response = await this.api.post('/api/faturas', { cpf });
        return response.data.message;
    }

    async listarFaturas(cpf) {
        const response = await this.api.get(`/api/faturas/${cpf}`);
        return response.data;
    }

    async obterEstatisticas() {
        const response = await this.api.get('/api/estatisticas');
        return response.data.message;
    }
}

function exibirMenu() {
    console.log('\n╔════════════════════════════════════╗');
    console.log('║     MENU PRINCIPAL (NODE.JS)       ║');
    console.log('╠════════════════════════════════════╣');
    console.log('║  1 - Adicionar Cliente             ║');
    console.log('║  2 - Remover Cliente               ║');
    console.log('║  3 - Consultar Cliente             ║');
    console.log('║  4 - Listar Clientes               ║');
    console.log('║  5 - Adicionar Linha               ║');
    console.log('║  6 - Remover Linha                 ║');
    console.log('║  7 - Registrar Chamada             ║');
    console.log('║  8 - Gerar Fatura                  ║');
    console.log('║  9 - Listar Faturas                ║');
    console.log('║ 10 - Ver Estatísticas              ║');
    console.log('║  0 - Sair                          ║');
    console.log('╚════════════════════════════════════╝');
}

async function pergunta(rl, texto) {
    return new Promise((resolve) => {
        rl.question(texto, (resposta) => resolve(resposta.trim()));
    });
}

async function main() {
    console.log('╔════════════════════════════════════════════════════════╗');
    console.log('║   CLIENTE NODE.JS - SISTEMA DE TELEFONIA REST API      ║');
    console.log('╚════════════════════════════════════════════════════════╝');

    const api = new ClienteTelefoniaAPI();
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    // Verificar conexão
    console.log('\n🔌 Verificando conexão com servidor...');
    if (!(await api.verificarServidor())) {
        console.log('❌ Erro: Servidor não está respondendo!');
        console.log('💡 Certifique-se de que o servidor Java está rodando na porta 8080');
        rl.close();
        return;
    }
    console.log('✅ Conectado ao servidor!');

    let continuar = true;

    while (continuar) {
        try {
            exibirMenu();
            const opcao = await pergunta(rl, 'Opção: ');

            switch (opcao) {
                case '1': {
                    console.log('\n=== ADICIONAR CLIENTE ===');
                    const nome = await pergunta(rl, 'Nome: ');
                    const cpf = await pergunta(rl, 'CPF: ');
                    const telefone = await pergunta(rl, 'Telefone: ');
                    const email = await pergunta(rl, 'Email: ');
                    const resultado = await api.adicionarCliente(nome, cpf, telefone, email);
                    console.log('\n📌', resultado);
                    break;
                }

                case '2': {
                    console.log('\n=== REMOVER CLIENTE ===');
                    const cpf = await pergunta(rl, 'CPF: ');
                    const resultado = await api.removerCliente(cpf);
                    console.log('\n📌', resultado);
                    break;
                }

                case '3': {
                    console.log('\n=== CONSULTAR CLIENTE ===');
                    const cpf = await pergunta(rl, 'CPF: ');
                    const resultado = await api.consultarCliente(cpf);
                    console.log('\n', resultado);
                    break;
                }

                case '4': {
                    console.log('\n=== LISTA DE CLIENTES ===');
                    const clientes = await api.listarClientes();
                    if (clientes.length > 0) {
                        clientes.forEach((cliente, i) => {
                            console.log(`${i + 1}. ${cliente}`);
                        });
                    } else {
                        console.log('Nenhum cliente cadastrado');
                    }
                    break;
                }

                case '5': {
                    console.log('\n=== ADICIONAR LINHA ===');
                    const cpf = await pergunta(rl, 'CPF do cliente: ');
                    const numero = await pergunta(rl, 'Número da linha: ');
                    const resultado = await api.adicionarLinha(cpf, numero);
                    console.log('\n📌', resultado);
                    break;
                }

                case '6': {
                    console.log('\n=== REMOVER LINHA ===');
                    const cpf = await pergunta(rl, 'CPF do cliente: ');
                    const numero = await pergunta(rl, 'Número da linha: ');
                    const resultado = await api.removerLinha(cpf, numero);
                    console.log('\n📌', resultado);
                    break;
                }

                case '7': {
                    console.log('\n=== REGISTRAR CHAMADA ===');
                    const origem = await pergunta(rl, 'Número de origem: ');
                    const destino = await pergunta(rl, 'Número de destino: ');
                    const duracao = parseInt(await pergunta(rl, 'Duração (minutos): '));
                    const resultado = await api.registrarChamada(origem, destino, duracao);
                    console.log('\n📌', resultado);
                    break;
                }

                case '8': {
                    console.log('\n=== GERAR FATURA ===');
                    const cpf = await pergunta(rl, 'CPF do cliente: ');
                    const resultado = await api.gerarFatura(cpf);
                    console.log('\n', resultado);
                    break;
                }

                case '9': {
                    console.log('\n=== LISTAR FATURAS ===');
                    const cpf = await pergunta(rl, 'CPF do cliente: ');
                    const faturas = await api.listarFaturas(cpf);
                    console.log();
                    faturas.forEach(fatura => console.log(fatura));
                    break;
                }

                case '10': {
                    console.log('\n=== ESTATÍSTICAS DO SISTEMA ===');
                    const resultado = await api.obterEstatisticas();
                    console.log('\n', resultado);
                    break;
                }

                case '0': {
                    console.log('\n👋 Encerrando cliente Node.js...');
                    continuar = false;
                    break;
                }

                default:
                    console.log('❌ Opção inválida!');
            }
        } catch (error) {
            console.error(`❌ Erro: ${error.message}`);
        }
    }

    rl.close();
}

main().catch(console.error);
